File: pi_host/pnm_pi.go
// pnm_pi.go — PNM orchestrator host driver for Raspberry Pi Compute Modules over SPI.
//
// Go alternative to pnm_pi.py (same wire protocol, same register API).
// Standard library only: talks to spidev through raw ioctl, no cgo, so it
// cross-compiles to the Pi with GOOS=linux GOARCH=arm64 from any host:
//
//	go build -o pnm_pi_go pnm_pi.go
//
// Targets CM0/CM1/CM3/CM4 class modules where pi_bridge (HDL/pi_bridge.v)
// is wired to a full SPI bus. Frame format (48 bits, MSB first, mode 0):
//
//	byte 0   : header {rw(1b) << 7 | sel[6:0]}   rw=1 read, rw=0 write
//	bytes 1-4: 32-bit data field, MSB first (sent on MOSI / returned on MISO)
//	byte 5   : status byte echoed back (0x01 write-ack, 0x00 read-ok)
package main

import (
	"encoding/binary"
	"flag"
	"fmt"
	"os"
	"runtime"
	"syscall"
	"unsafe"
)

const (
	selCtrl       = 0x00
	selLayer      = 0x01
	selModule     = 0x02
	selLen        = 0x03
	selData       = 0x04
	selStatus     = 0x05
	selResult     = 0x06
	selErrors     = 0x07
	selDispatches = 0x08
	selWeights    = 0x09

	ctrlInject   = 1 << 0
	ctrlBootDone = 1 << 2

	statusBusy = 1 << 0

	frameLen = 6
)

// Linux ioctl encoding (asm-generic): 14-bit size, 8-bit type, 8-bit nr,
// 2-bit dir. Matches linux/spi/spidev.h on all Pi targets.
const (
	iocNrbits   = 8
	iocTypebits = 8
	iocSizebits = 14

	iocNrshift   = 0
	iocTypeshift = iocNrshift + iocNrbits
	iocSizeshift = iocTypeshift + iocTypebits
	iocDirshift  = iocSizeshift + iocSizebits

	iocWrite = 1
	iocRead  = 2

	spiIocMagic = 'k'
)

func ioc(dir, typ, nr, size uint) uint {
	return dir<<iocDirshift | typ<<iocTypeshift | nr<<iocNrshift | size<<iocSizeshift
}

// spiIocTransfer mirrors struct spi_ioc_transfer (32 bytes, kernel ABI).
type spiIocTransfer struct {
	txBuf       uint64
	rxBuf       uint64
	length      uint32
	speedHz     uint32
	delayUsecs  uint16
	bitsPerWord uint8
	csChange    uint8
	txNBits     uint8
	rxNBits     uint8
	pad         uint16
}

const spiIocTransferSize = uint(unsafe.Sizeof(spiIocTransfer{}))

func spiIocMessage(n int) uint {
	size := uint(n) * spiIocTransferSize
	const max = 1 << iocSizebits
	if size > max {
		size = max
	}
	return ioc(iocWrite, spiIocMagic, 0, size)
}

var (
	spiIocWrMode        = ioc(iocWrite, spiIocMagic, 1, 1)
	spiIocWrBitsPerWord = ioc(iocWrite, spiIocMagic, 3, 1)
	spiIocWrMaxSpeedHz  = ioc(iocWrite, spiIocMagic, 4, 4)
)

// PNM is a host-side handle for one pi_bridge SPI device.
type PNM struct {
	f     *os.File
	speed uint32
}

// Open claims /dev/spidev<bus>.<dev> and configures mode 0 @ speedHz.
func Open(bus, dev int, speedHz uint32) (*PNM, error) {
	path := fmt.Sprintf("/dev/spidev%d.%d", bus, dev)
	f, err := os.OpenFile(path, os.O_RDWR, 0)
	if err != nil {
		return nil, err
	}
	p := &PNM{f: f, speed: speedHz}
	mode := uint8(0) // SPI_MODE_0
	bits := uint8(8)
	for _, c := range []struct {
		cmd uint
		val unsafe.Pointer
	}{{spiIocWrMode, unsafe.Pointer(&mode)},
		{spiIocWrBitsPerWord, unsafe.Pointer(&bits)},
		{spiIocWrMaxSpeedHz, unsafe.Pointer(&p.speed)}} {
		if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, f.Fd(),
			uintptr(c.cmd), uintptr(c.val)); errno != 0 {
			f.Close()
			return nil, errno
		}
	}
	return p, nil
}

// Close releases the SPI device.
func (p *PNM) Close() error { return p.f.Close() }

// xfer runs one full-duplex 48-bit frame and returns the MISO bytes.
func (p *PNM) xfer(frame []byte) ([]byte, error) {
	resp := make([]byte, frameLen)
	tr := spiIocTransfer{
		txBuf:       uint64(uintptr(unsafe.Pointer(&frame[0]))),
		rxBuf:       uint64(uintptr(unsafe.Pointer(&resp[0]))),
		length:      frameLen,
		speedHz:     p.speed,
		bitsPerWord: 8,
	}
	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, p.f.Fd(),
		uintptr(spiIocMessage(1)), uintptr(unsafe.Pointer(&tr)))
	runtime.KeepAlive(frame)
	runtime.KeepAlive(resp)
	runtime.KeepAlive(tr)
	if errno != 0 {
		return nil, errno
	}
	return resp, nil
}

func (p *PNM) frame(rw bool, sel byte, data uint32) []byte {
	fr := make([]byte, frameLen)
	hdr := byte(sel & 0x7F)
	if rw {
		hdr |= 0x80
	}
	fr[0] = hdr
	binary.BigEndian.PutUint32(fr[1:5], data)
	return fr
}

// RegWrite stores value into register sel; errors on bad ack status.
func (p *PNM) RegWrite(sel byte, value uint32) error {
	resp, err := p.xfer(p.frame(false, sel, value))
	if err != nil {
		return err
	}
	if resp[5] != 0x01 {
		return fmt.Errorf("pi_bridge write-ack status 0x%02x", resp[5])
	}
	return nil
}

// RegRead fetches the current value of register sel.
func (p *PNM) RegRead(sel byte) (uint32, error) {
	resp, err := p.xfer(p.frame(true, sel, 0))
	if err != nil {
		return 0, err
	}
	if resp[5] != 0x00 {
		return 0, fmt.Errorf("pi_bridge read status 0x%02x", resp[5])
	}
	return binary.BigEndian.Uint32(resp[1:5]), nil
}

// BootDone signals the orchestrator that firmware boot completed.
func (p *PNM) BootDone() error { return p.RegWrite(selCtrl, ctrlBootDone) }

// DispatchCount reports the chip's dispatch counter.
func (p *PNM) DispatchCount() (uint32, error) { return p.RegRead(selDispatches) }

// Inject streams one flit into the chip (busy-waits while STATUS_BUSY).
func (p *PNM) Inject(layer, module byte, payload []byte) error {
	for {
		st, err := p.RegRead(selStatus)
		if err != nil {
			return err
		}
		if st&statusBusy == 0 {
			break
		}
	}
	if err := p.RegWrite(selLayer, uint32(layer)); err != nil {
		return err
	}
	if err := p.RegWrite(selModule, uint32(module)); err != nil {
		return err
	}
	if err := p.RegWrite(selLen, uint32(len(payload))); err != nil {
		return err
	}
	for _, b := range payload {
		if err := p.RegWrite(selData, uint32(b)); err != nil {
			return err
		}
	}
	return p.RegWrite(selCtrl, ctrlInject)
}

func main() {
	bus := flag.Int("bus", 0, "spidev bus number")
	dev := flag.Int("dev", 0, "spidev device number")
	speed := flag.Int("speed", 10_000_000, "SCLK rate in Hz")
	flag.Parse()

	pnm, err := Open(*bus, *dev, uint32(*speed))
	if err != nil {
		fmt.Fprintln(os.Stderr, "open:", err)
		os.Exit(1)
	}
	defer pnm.Close()

	const magic = uint32(0xDEADBEEF)
	if err := pnm.RegWrite(selLayer, magic); err != nil {
		fmt.Fprintln(os.Stderr, "write:", err)
		os.Exit(1)
	}
	got, err := pnm.RegRead(selLayer)
	if err != nil {
		fmt.Fprintln(os.Stderr, "read:", err)
		os.Exit(1)
	}
	status := "OK"
	if got != magic {
		status = "FAIL"
	}
	fmt.Printf("layer round-trip: wrote 0x%08x read 0x%08x -> %s\n",
		magic, got, status)
	n, _ := pnm.DispatchCount()
	fmt.Println("dispatches so far:", n)
}

File: pi_host/pnm_pi.py
#!/usr/bin/env python3
"""pnm_pi.py — PNM router host driver for Raspberry Pi Compute Modules over SPI.

Targets CM0/CM1/CM3/CM4 class modules where the PNM pi_bridge is wired to a
full SPI bus (spidev). For CM5 prefer pnm_pi_cm5.c (PCIe BAR access); this
module still works there through the SPI pins if wired.

Frame format (48 bits, matches HDL/pi_bridge.v):
    byte 0   : header {rw(1b) << 7 | sel[6:0]}     rw=1 read, rw=0 write
    bytes 1-4: 32-bit data field, MSB first
               write: value sent to the bridge
               read : register value returned by the bridge on MISO
    byte 5   : status byte echoed back by the bridge (0x01 write-ack,
               0x00 read-ok)

Requires: spidev (kernel module spi_bcm2835 enabled, spidev device present).
The bridge needs the system clock >= ~10x SCLK; keep max_speed_hz at or
below what the FPGA side can settle.
"""

import time

SPI_MODE_0 = 0

SEL_CTRL = 0x00
SEL_LAYER = 0x01
SEL_MODULE = 0x02
SEL_LEN = 0x03
SEL_DATA = 0x04
SEL_STATUS = 0x05
SEL_RESULT = 0x06
SEL_ERRORS = 0x07
SEL_DISPATCHES = 0x08
SEL_WEIGHTS = 0x09

CTRL_INJECT = 1 << 0
CTRL_BOOTDONE = 1 << 2

STATUS_BUSY = 1 << 0


class PNMPi:
    """Host-side register access to a PNM chassis through pi_bridge."""

    def __init__(self, bus=0, device=0, speed_hz=10_000_000):
        import spidev

        self._spi = spidev.SpiDev()
        self._spi.open(bus, device)
        self._spi.mode = SPI_MODE_0
        self._spi.max_speed_hz = speed_hz
        self._spi.bits_per_word = 8

    def close(self):
        self._spi.close()

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

    def _frame(self, rw, sel, data=0):
        hdr = ((rw & 1) << 7) | (sel & 0x7F)
        return bytes([hdr]) + data.to_bytes(4, "big") + b"\x00"

    def reg_write(self, sel, value):
        resp = self._spi.xfer2(self._frame(0, sel, value & 0xFFFFFFFF))
        status = resp[5]
        if status != 0x01:
            raise IOError("pi_bridge write-ack status 0x%02x" % status)

    def reg_read(self, sel):
        resp = self._spi.xfer2(self._frame(1, sel))
        if resp[5] != 0x00:
            raise IOError("pi_bridge read status 0x%02x" % resp[5])
        return int.from_bytes(resp[1:5], "big")

    def boot_done(self):
        self.reg_write(SEL_CTRL, CTRL_BOOTDONE)

    def dispatch_count(self):
        return self.reg_read(SEL_DISPATCHES)

    def inject(self, layer, module, payload):
        while self.reg_read(SEL_STATUS) & STATUS_BUSY:
            time.sleep(0)
        self.reg_write(SEL_LAYER, layer & 0xFF)
        self.reg_write(SEL_MODULE, module & 0xFF)
        self.reg_write(SEL_LEN, len(payload))
        for b in payload:
            self.reg_write(SEL_DATA, b)
        self.reg_write(SEL_CTRL, CTRL_INJECT)


def main():
    import argparse

    ap = argparse.ArgumentParser(description="PNM pi_bridge smoke test")
    ap.add_argument("--bus", type=int, default=0)
    ap.add_argument("--dev", type=int, default=0)
    ap.add_argument("--speed", type=int, default=10_000_000)
    args = ap.parse_args()

    with PNMPi(args.bus, args.dev, args.speed) as pnm:
        magic = 0xDEADBEEF
        pnm.reg_write(SEL_LAYER, magic)
        got = pnm.reg_read(SEL_LAYER)
        print("layer round-trip: wrote 0x%08x read 0x%08x -> %s"
              % (magic, got, "OK" if got == magic else "FAIL"))
        print("dispatches so far:", pnm.dispatch_count())


if __name__ == "__main__":
    main()
